### DISCUSSION TITLE
How to access the hex request and response in a pre processor

### DISCUSSION THREAD

**Discussion by AlphaNiner1415** [tilln/jmeter-iso8583#85] [Q&A]
Hi, so I tried to access the request/response of the sampler in a Beanshell Preprocessor with sampler.getSamplerData() and it is saying this method doesn't exist. So how can I get the request/response from the sampler?

**Comment by tilln** [comment]
Hi @AlphaNiner1415 

> I tried to access the request/response of the sampler in a Beanshell Preprocessor

In a Pre-Processor only the request can be accessed as the response is unknown until *after* the Sampler has run.
You may want to use a Post-Processor for the response.
Also, I would recommend using a JSR223 Post-Processor instead of Beanshell. See [here](https://jmeter.apache.org/usermanual/best-practices.html#jsr223) and [here](https://www.blazemeter.com/blog/beanshell-vs-jsr223-vs-jmeter).

> sampler.getSamplerData() and it is saying this method doesn't exist

Not the Sampler but the [SampleResult](https://jmeter.apache.org/api/org/apache/jmeter/samplers/SampleResult.html#getSamplerData()) provides this method. Also not defined in a Pre-Processor.

> So how can I get the request/response from the sampler?

1. In a Pre-Processor or Post-Processor:
    [`sampler.getRequest()`](https://github.com/tilln/jmeter-iso8583/blob/1.3/src/main/java/nz/co/breakpoint/jmeter/iso8583/ISO8583Sampler.java#L233) will create and return the same [`ISOMsg`](https://jpos.org/doc/javadoc/org/jpos/iso/ISOMsg.html) object as the sampler will send.
This will include fields added by an earlier Crypto PreProcessor (MAC, ARQC, etc.).
For example, if you want the message content as a hex string:
`org.jpos.iso.ISOUtil.byte2hex(sampler.getRequest().pack())`

2. In a Post-Processor:
    [`sampler.getResponse()`](https://github.com/tilln/jmeter-iso8583/blob/1.3/src/main/java/nz/co/breakpoint/jmeter/iso8583/ISO8583Sampler.java#L234) will return the response `ISOMsg` the sampler received which you can turn into a hex string as above.
    The XML representation you would see in a "View Results Tree" can be accessed via `prev.getSamplerData()` for the request and `prev.getResponseDataAsString()` for the response.
    Note `prev` is the SampleResult object in a BeanShell or JSR223 Post-Processor.
